File: deploy_local_with_migration.py
#!/usr/local/bin/python3

import json
import os
from deploy_local import *
from terra_sdk.core.wasm import MsgMigrateContract
import sys

TERRA_ORIGINAL_HUB_CONTRACT = "anchor_basset_hub.wasm"
TERRA_ORIGINAL_REWARD_CONTRACT = "anchor_basset_reward.wasm"
TERRA_ORIGINAL_TOKEN_CONTRACT = "anchor_basset_token.wasm"


def migrate_contract(terra_client, wallet, contract_address, new_code_id, message):
    print("Migrating {0} to new code_id = {1}...".format(contract_address, new_code_id))

    execute = MsgMigrateContract(
        wallet.key.acc_address,
        contract_address,
        new_code_id,
        message,
    )

    migrate_tx = wallet.create_and_sign_tx(
        msgs=[execute], fee=StdFee(1000000, Coins(uluna=1000000))
    )

    migrate_tx_result = terra_client.tx.broadcast(migrate_tx)
    if migrate_tx_result.code is not None:
        print("Error while migrating contract: {}".format(migrate_tx_result.raw_log))
        exit(1)

    return migrate_tx_result


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Please provide a path to the original Terra contract's folder")
        exit(1)

    terra = LocalTerra()
    test1 = terra.wallets["test1"]

    print("Uploading original Terra contracts...")

    TERRA_ORIGINAL_CONTRACTS_FOLDER = sys.argv[1]

    original_hub_code_id = store_code(terra, test1, os.path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER,
                                                                 TERRA_ORIGINAL_HUB_CONTRACT))
    original_reward_code_id = store_code(terra, test1, os.path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER,
                                                                    TERRA_ORIGINAL_REWARD_CONTRACT))
    original_bluna_token_code_id = store_code(terra, test1, os.path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER,
                                                                         TERRA_ORIGINAL_TOKEN_CONTRACT))

    hub_address = instantiate_contract(terra, test1, original_hub_code_id,
                                       {"epoch_period": 30, "er_threshold": "10000000000000", "peg_recovery_fee": "0",
                                        "reward_denom": "uusd", "unbonding_period": 2, "underlying_coin_denom": "uluna",
                                        "validator": "terravaloper1dcegyrekltswvyy0xy69ydgxn9x8x32zdy3ua5"},
                                       Coins(uluna=1000000))

    reward_address = instantiate_contract(terra, test1, original_reward_code_id,
                                          {"hub_contract": hub_address, "reward_denom": "uusd"})

    bluna_token_address = instantiate_contract(terra, test1, original_bluna_token_code_id,
                                               {"decimals": 6, "hub_contract": hub_address,
                                                "initial_balances": [{"address": hub_address, "amount": "1000000"}],
                                                "name": "bluna", "symbol": "BLUNA",
                                                "mint": {"minter": hub_address, "cap": None}})

    update_hub_config = execute_contract(terra, test1, hub_address, {
        "update_config": {"token_contract": bluna_token_address,
                          "reward_contract": reward_address}})

    print()
    print("Starting migration process...")

    new_hub_code_id = store_code(terra, test1, "artifacts/anchor_basset_hub.wasm")
    new_reward_code_id = store_code(terra, test1, "artifacts/anchor_basset_reward.wasm")
    new_bluna_token_code_id = store_code(terra, test1, "artifacts/anchor_basset_token.wasm")
    rewards_dispatcher_code_id = store_code(terra, test1, "artifacts/anchor_basset_rewards_dispatcher.wasm")
    validators_registry_code_id = store_code(terra, test1, "artifacts/anchor_basset_validators_registry.wasm")
    stluna_token_code_id = store_code(terra, test1, "artifacts/anchor_basset_token_stluna.wasm")

    rewards_dispatcher_address = instantiate_contract(terra, test1, rewards_dispatcher_code_id,
                                                      {"lido_fee_address": test1.key.acc_address,
                                                       "lido_fee_rate": "0.05", "hub_contract": hub_address,
                                                       "bluna_reward_contract": reward_address,
                                                       "stluna_reward_denom": "uluna", "bluna_reward_denom": "uusd"})

    validators_registry_address = instantiate_contract(terra, test1, validators_registry_code_id,
                                                       {"hub_contract": hub_address,
                                                        "registry": []})

    stluna_token_address = instantiate_contract(terra, test1, stluna_token_code_id,
                                                {"decimals": 6, "hub_contract": hub_address, "initial_balances": [],
                                                 "name": "stluna", "symbol": "STLUNA",
                                                 "mint": {"minter": hub_address, "cap": None}})

    migrate_hub = migrate_contract(terra, test1, hub_address, new_hub_code_id, {
        "stluna_exchange_rate": "1.0",
        "total_bond_stluna_amount": "0",
        "reward_dispatcher_contract": rewards_dispatcher_address,
        "validators_registry_contract": validators_registry_address,
        "stluna_token_contract": stluna_token_address
    })
    migrate_rewards = migrate_contract(terra, test1, reward_address, new_reward_code_id, {})
    migrate_bluna_token = migrate_contract(terra, test1, bluna_token_address, new_bluna_token_code_id, {})

    print()
    print("HUB_CONTRACT = {}".format(hub_address))
    print("REWARD_CONTRACT = {}".format(reward_address))
    print("REWARDS_DISPATCHER_CONTRACT = {}".format(rewards_dispatcher_address))
    print("VALIDATORS_REGISTRY_CONTRACT = {}".format(validators_registry_address))
    print("BLUNA_TOKEN_CONTRACT = {}".format(bluna_token_address))
    print("STLUNA_TOKEN_CONTRACT = {}".format(stluna_token_address))

    # just a few tests to make sure the contracts works
    # for more accurate tests we must use integration-tests repo
    execute_contract(terra, test1, hub_address, {"bond_for_st_luna": {}}, Coins(uluna=1000000))
    execute_contract(terra, test1, hub_address, {"bond": {}}, Coins(uluna=1000000))

    execute_contract(terra, test1, hub_address, {"bond_for_st_luna": {}}, Coins(uluna=1000000))
    execute_contract(terra, test1, hub_address, {"bond": {}}, Coins(uluna=1000000))

    execute_contract(terra, test1, stluna_token_address, {"send": {"contract": hub_address, "amount": "1000000",
                                                                   "msg": base64.b64encode(
                                                                       json.dumps({"unbond": {}}).encode(
                                                                           'utf-8')).decode('utf-8')}})
    execute_contract(terra, test1, bluna_token_address, {"send": {"contract": hub_address, "amount": "1000000",
                                                                  "msg": base64.b64encode(
                                                                      json.dumps({"unbond": {}}).encode(
                                                                          'utf-8')).decode('utf-8')}})

File: deployment_scripts/deploy_local_with_migration.ts
import {Coins, Key, LocalTerra, MnemonicKey, MsgSend} from "@terra-money/terra.js";
import {executeContract, instantiateContract, migrateContract, storeCode} from "./common";
const path = require('path');

const TERRA_ORIGINAL_HUB_CONTRACT = "lido_terra_hub.wasm"
const TERRA_ORIGINAL_REWARD_CONTRACT = "lido_terra_reward.wasm"
const TERRA_ORIGINAL_TOKEN_CONTRACT = "lido_terra_token.wasm"

async function main(): Promise<void> {
  if (process.argv.length < 3) {
    throw new Error("Provide path to the original Terra contracts, please. ts-node deploy_local_with_migration.ts ~/original_contracts")
  }
  const terra = new LocalTerra();
  const {test1} = terra.wallets;

  console.log("Uploading original Terra contracts...")

  const TERRA_ORIGINAL_CONTRACTS_FOLDER = process.argv[2]

  let originalHubCodeId = await storeCode(terra, test1, path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER, TERRA_ORIGINAL_HUB_CONTRACT))
  let originalRewardCodeId = await storeCode(terra, test1, path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER, TERRA_ORIGINAL_REWARD_CONTRACT))
  let originalBlunaTokenCodeId = await storeCode(terra, test1, path.join(TERRA_ORIGINAL_CONTRACTS_FOLDER, TERRA_ORIGINAL_TOKEN_CONTRACT))

  let hubAddress = await instantiateContract(terra, test1, originalHubCodeId,
    {epoch_period: 300, er_threshold: "1.0", peg_recovery_fee: "0", reward_denom: "uusd", unbonding_period: 6, underlying_coin_denom: "uluna", validator: "terravaloper1dcegyrekltswvyy0xy69ydgxn9x8x32zdy3ua5"}, new Coins({uluna: 1000000}))

  let rewardAddress = await instantiateContract(terra, test1, originalRewardCodeId,
    {hub_contract: hubAddress, reward_denom: "uusd"}, new Coins({}))

  let blunaTokenAddress = await instantiateContract(terra, test1, originalBlunaTokenCodeId,
    {decimals: 6, hub_contract: hubAddress, initial_balances: [{address: hubAddress, amount: "1000000"}],
      name: "bluna", symbol: "BLUNA",
      mint: {minter: hubAddress, cap: null}}, new Coins({}))

  await executeContract(terra, test1, hubAddress, {
    update_config: {token_contract: blunaTokenAddress, reward_contract: rewardAddress}}, new Coins({}))

  let wallets = [];
  // simulate large unbond queue from many users
  for (let i = 0; i < 3000; i++ ) {
    let wallet = terra.wallet(new MnemonicKey());
    const send = await test1.createAndSignTx({msgs: [new MsgSend(
      test1.key.accAddress,
      wallet.key.accAddress,
      { uluna: 1000000000, uusd: 1000000000}
    )]});
    await terra.tx.broadcast(send);

    await executeContract(terra, wallet, hubAddress, {bond: {validator: "terravaloper1dcegyrekltswvyy0xy69ydgxn9x8x32zdy3ua5"}}, new Coins({uluna: 1000000000}))

    await executeContract(terra, wallet, blunaTokenAddress, {send: {contract: hubAddress, amount: "100",
        msg: Buffer.from(JSON.stringify({"unbond": {}})).toString('base64')}}, new Coins({}));

    console.log("UNBOND REQUEST NUMBER #", i, "from address ", wallet.key.accAddress);

    wallets.push(wallet);
  }

  console.log()
  console.log("Starting migration process...")

  let newHubCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_hub.wasm")
  let newRewardCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_reward.wasm")
  let newBlunaTokenCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_token.wasm")
  let rewardsDispatcherCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_rewards_dispatcher.wasm")
  let validatorsRegistryCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_validators_registry.wasm")
  let stlunaTokenCodeId = await storeCode(terra, test1, "../artifacts/lido_terra_token_stluna.wasm")

  let rewardsDispatcherAddress = await instantiateContract(terra, test1, rewardsDispatcherCodeId,
    {lido_fee_address: test1.key.accAddress,
      lido_fee_rate: "0.05", hub_contract: hubAddress, bluna_reward_contract: rewardAddress,
      stluna_reward_denom: "uluna", bluna_reward_denom: "uusd"}, new Coins({}))

  let validatorsRegistryAddress = await instantiateContract(terra, test1, validatorsRegistryCodeId,
    {hub_contract: hubAddress,
      registry: [{active: true,
        address: "terravaloper1dcegyrekltswvyy0xy69ydgxn9x8x32zdy3ua5",
        total_delegated: "0"}]}, new Coins({}))

  let stlunaTokenAddress = await instantiateContract(terra, test1, stlunaTokenCodeId,
    {decimals: 6, hub_contract: hubAddress, initial_balances: [],
      name: "stluna", symbol: "STLUNA",
      mint: {minter: hubAddress, cap: null}}, new Coins({}))

  console.log("Migrating hub...")
  await migrateContract(terra, test1, hubAddress, newHubCodeId, {
    reward_dispatcher_contract: rewardsDispatcherAddress,
    validators_registry_contract: validatorsRegistryAddress,
    stluna_token_contract: stlunaTokenAddress
  })
  console.log("Done")

  try {
    await executeContract(terra, test1, hubAddress, {withdraw_unbonded: {}}, new Coins({}));
  } catch (e) {
    // the hub is paused
    console.log("Error: ", e.response.data.error)
  }

  console.log("Migrating rewards...")
  await migrateContract(terra, test1, rewardAddress, newRewardCodeId, {})
  console.log("Done")

  console.log("Migrating bLuna token...")
  await migrateContract(terra, test1, blunaTokenAddress, newBlunaTokenCodeId, {})
  console.log("Done")

  for (let i = 0; i < 4; i++ ) {
    try {
      await executeContract(terra, test1, hubAddress, {
        update_params: {paused: false}}, new Coins({}));
    } catch (e) {
      // cannot unpause the hub with old unbond wait lists
      console.log("Error: ", e.response.data.error)
    }
    let response = await executeContract(terra, test1, hubAddress, {migrate_unbond_wait_list: {limit: 1000}}, new Coins({}));
    console.log(response.raw_log);
  }

  console.log()

  console.log(`HUB_CONTRACT = ${hubAddress}`)
  console.log(`REWARD_CONTRACT = ${rewardAddress}`)
  console.log(`REWARDS_DISPATCHER_CONTRACT = ${rewardsDispatcherAddress}`)
  console.log(`VALIDATORS_REGISTRY_CONTRACT = ${validatorsRegistryAddress}"`)
  console.log(`BLUNA_TOKEN_CONTRACT = ${blunaTokenAddress}`)
  console.log(`STLUNA_TOKEN_CONTRACT = ${stlunaTokenAddress}`)

  await new Promise(r => setTimeout(r, 10000));

  await executeContract(terra, test1, hubAddress, {bond: {validator: "terravaloper1dcegyrekltswvyy0xy69ydgxn9x8x32zdy3ua5"}}, new Coins({uluna: 1000000000}))
  await executeContract(terra, test1, blunaTokenAddress, {send: {contract: hubAddress, amount: "100",
      msg: Buffer.from(JSON.stringify({"unbond": {}})).toString('base64')}}, new Coins({}));


  //just a few simple tests to make sure the contracts are not failing
  //for more accurate tests we must use integration-tests repo
  for (let i = 0; i < 100; i++ ) {
    await executeContract(terra, wallets[i], hubAddress, {withdraw_unbonded: {}}, new Coins({}));
  }

  await executeContract(terra, test1, hubAddress, {bond_for_st_luna: {}}, new Coins({uluna: 1000000}))
  await executeContract(terra, test1, hubAddress, {bond: {}}, new Coins({uluna: 1000000}))

  await executeContract(terra, test1, hubAddress, {bond_for_st_luna: {}}, new Coins({uluna: 1000000}))
  await executeContract(terra, test1, hubAddress, {bond: {}}, new Coins({uluna: 1000000}))

  await executeContract(terra, test1, stlunaTokenAddress, {send: {contract: hubAddress, amount: "1000000",
      msg: Buffer.from(JSON.stringify({"unbond": {}})).toString('base64')}}, new Coins({}))

  await executeContract(terra, test1, blunaTokenAddress, {send: {contract: hubAddress, amount: "1000000",
      msg: Buffer.from(JSON.stringify({"unbond": {}})).toString('base64')}}, new Coins({}))
}

main().catch(console.log);
